shared/route-tree/render-route.js

```diff
@@ -0,0 +1,161 @@
+// @flow
+import * as I from 'immutable'
+import React, {PureComponent} from 'react'
+import {LeafTags, pathToString} from './'
+
+import type {RouteDefNode, RouteStateNode} from './'
+
+// Components rendered by routes receive the following props:
+export type RouteProps<P, S> = {
+  // Route props (query params)
+  routeProps: P,
+
+  // Route state (state associated with this path. can change, see below)
+  routeState: S,
+
+  // The name of the selected child route (useful for navs)
+  routeSelected: string,
+
+  // The path leading up to this component.
+  routePath: I.List<string>,
+
+  // Leaf tags for the current leaf route (e.g. is the child modal?)
+  routeLeafTags: LeafTags,
+
+  // Stack of child views from this component.
+  routeStack: I.List<React$Element<any>>,
+
+  // Call to update the state of the route node that rendered this component.
+  setRouteState: (partialState: {}) => void,
+}
+
+type RenderRouteNodeProps = {
+  isContainer: boolean,
+  routeDef: RouteDefNode,
+  routeState: RouteStateNode,
+  setRouteState: (partialState: {}) => void,
+  path: I.List<string>,
+  leafTags?: LeafTags,
+  stack?: I.List<React$Element<any>>,
+  children?: React$Element<*>,
+}
+
+// Helper to render a component based on route state and use
+// shouldComponentUpdate (via PureComponent).
+class RenderRouteNode extends PureComponent<*, RenderRouteNodeProps, *> {
+  render () {
+    const {isContainer, routeDef, routeState, setRouteState, path, leafTags, stack, children} = this.props
+    const RouteComponent = isContainer ? routeDef.containerComponent : routeDef.component
+    return (
+      <RouteComponent
+        routeProps={routeState.props.toJS()}
```

Comment:
Great point, this should use `.toObject()`. Record type could work. I initially wanted to use records for this but didn't want to have to pass the constructors around for each different routeProp type. Now that I'm thinking about it again, could probably just construct once in the routeDef and mutate it to derive changed copies. This could produce better types, though I've found flow to be a bit cumbersome with them since the `Record` type def currently only covers the set/get/delete functions, not the property getters or setters.